---
title: "TSScallingFromCappableSeq-data"
author: "Fabian Blombach"
date: "02/12/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

wt.r1.fw <- "../data/Cappable-seq_S.islandicus/E234_r1.5primeCov.plus.bw"
wt.r1.rv <- "../data/Cappable-seq_S.islandicus/E234_r1.5primeCov.minus.bw"
wt.r2.fw <- "../data/Cappable-seq_S.islandicus/E234_r2.5primeCov.plus.bw"
wt.r2.rv <- "../data/Cappable-seq_S.islandicus/E234_r2.5primeCov.minus.bw"

ko.r1.fw <- "../data/Cappable-seq_S.islandicus/cbp1_r1.5primeCov.plus.bw"
ko.r1.rv <- "../data/Cappable-seq_S.islandicus/cbp1_r1.5primeCov.minus.bw"
ko.r2.fw <- "../data/Cappable-seq_S.islandicus/cbp1_r2.5primeCov.plus.bw"
ko.r2.rv <- "../data/Cappable-seq_S.islandicus/cbp1_r2.5primeCov.minus.bw"

#CID coordinates
cid <- read.table("../genome_data/SislREY15A/CIDs.bed")
#correction for shift of CID border by 817 bp  to match first AluI restriction site

#CRISPR array coordinates
crispr <- read.table("../genome_data/SislREY15A/crispr_loci.SislREY15A.bed")

#CRISPR leader TSS positions based on manual data inspection of cappable-seq data
crispr_leader <- data.frame(V1=rep("NC_017276.1",2), V2=c(732605,736615), V3=c(732615,736625), V4=c("locus_id2", "locus_id3"), V5=rep(0,2), V6=c("-","+"))  
#summit position of non-CRISPR Cbp1 binding sites
#nonCRISPR <- read.table("../data/SislCbp1_pooled_idrFilteredPeaks_summits.bed") #all peaks
nonCRISPR <- read.table("../data/SislCbp1_pooled_IDRfilteredPeaks_summits.min5Enr.bed") #peaks with minimum 5x enrichment
#nonCRISPR <- read.table("../data/MEME_motif_coordinates/motifStart_coordinates_SislREY15A.bed") #associated Cbp1 binding motifs 


#S. islandicus REY15A genome annotation
sisl.genome <- read.delim("../genome_data/SislREY15A/GCF_000189555.1_ASM18955v1_genomic.gff", skip=7, header=F)
sisl.genome <- sisl.genome[-c(1, nrow(sisl.genome)),] #remove first and last rows of gff file without gene information

#operon predition by Operon-mapper (Taboada et al. 2018, Bioinformatics)
operons <- read.csv("../genome_data/SislREY15A/Operon-mapper.output.csv", header=T)

require(rtracklayer)
require(DESeq2)
require(ggplot2)
require(RVAideMemoire)
require(LSD)
require(circlize)
require(dplyr)
```


```{r TSS candidates}
bigWigFiles<-c(wt.r1.fw, wt.r2.fw, ko.r1.fw, ko.r2.fw)
bigWigFiles.rv<-c(wt.r1.rv, wt.r2.rv, ko.r1.rv, ko.r2.rv)

TSS <- data.frame(matrix(ncol=4, nrow=2*2522992))
colnames(TSS) <- c("wt.r1", "wt.r2", "ko.r1", "ko.r2")
c.incr<- 20 #cut-off values for total and relative increase of signal compared to preceding position
c.ratio<- 5

for(i in 1:4){
  x<-unlist(import.bw(bigWigFiles[i], as = "NumericList")[[1]])
  x.rv<-unlist(import.bw(bigWigFiles.rv[i], as = "NumericList")[[1]])
  y<-vector(mode = "logical", length = length(x))
  y.rv<-vector(mode = "logical", length = length(x))
  #loop for fw bigwig file
  for (j in 2:length(x)){
    if(x[j] - x[j-1] >= c.incr & x[j] / x[j-1] >= c.ratio){
      y[j] <- T
    }
  }
  #loop for fw bigwig file
  for (k in 1:(length(x)-1)){
    if(x.rv[k] - x.rv[k+1] >= c.incr & x.rv[k] / x.rv[k+1] >= c.ratio){
      y.rv[k] <- T
    }
  }
TSS[,i]<- c(y, y.rv)
}
summary(TSS)


colnames(TSS)<- c("wt.r1", "wt.r2", "ko.r1", "ko.r2")
TSS$genomePos<-rep(1:2522992,2)
TSS$strand<-c(rep("+", 2522992), rep("-", 2522992))

cor(TSS[,1:4])

TSS$n.cons <- rowSums(TSS[,1:4])
summary(as.factor(TSS$n.cons))

TSS$n.cons.wt <- rowSums(TSS[,1:2])
summary(as.factor(TSS$n.cons.wt))
TSS$n.cons.ko <- rowSums(TSS[,3:4])
summary(as.factor(TSS$n.cons.ko))

```


```{r clustering close TSSs}
#call TSSs consistent between replicates of at least one condition
TSS.called <- as.data.frame(TSS$genomePos[TSS$n.cons >= 2])
colnames(TSS.called)<-"genomePos"
TSS.called$strand <- TSS$strand[TSS$n.cons >= 2]
#filter out TSSs within the cbp1 gene
TSS.called.filt<- TSS.called[!(TSS.called$genomePos %in% 1440640:1441370), ]


#join positions clusterering together within 5 bp distance
  #calculating initial TSS distance
distance.TSS<- vector() 
for(i in 2:nrow(TSS.called.filt)){
  distance.TSS[i] <- abs(TSS.called.filt$genomePos[i] - TSS.called.filt$genomePos[i-1])
}

distance.TSS.cur<-distance.TSS
TSS.called.init<-TSS.called.filt
repeat {
#cluster TSSs closer than 5bp distance together
TSS.called.cur<-data.frame(matrix(ncol=2, nrow=nrow(TSS.called.init)))
colnames(TSS.called.cur) <- colnames(TSS.called.init)
for(i in 1:length(distance.TSS.cur)){
  if(is.na(distance.TSS.cur)[i]==F & distance.TSS.cur[i] <= 5)
  {x <- c(ceiling(mean(TSS.called.init[(i-1):i,1])), TSS.called.init[i,2])
    TSS.called.cur[i,] <- x}
  else {if(i < length(distance.TSS.cur) & distance.TSS.cur[i+1] <= 5)
  {x <- c(ceiling(mean(TSS.called.init[i:(i+1),1])), TSS.called.init[i,2])
    TSS.called.cur[i,]<- x}
  else {TSS.called.cur[i,]<-TSS.called.init[i,]}
  }
}
TSS.called.init<-unique(TSS.called.cur) 
TSS.called.init$genomePos<-as.numeric(TSS.called.init$genomePos)
distance.TSS.cur <- vector() 
for(i in 2:nrow(TSS.called.init)){
  distance.TSS.cur[i] <- abs(TSS.called.init$genomePos[i] - TSS.called.init$genomePos[i-1])
}
#report numbers of TSSs still within cluster distance of 5 bp
print(sum(distance.TSS.cur[!is.na(distance.TSS.cur)] <= 5))
if (min(distance.TSS.cur[!is.na(distance.TSS.cur)]) > 5){
  TSS.called.final<-TSS.called.init
  break
  }
}
```


```{r count signal over 11 bp window}
countData <- matrix(1:(4*nrow(TSS.called.final)),ncol=4)
for(i in 1:4){
  x.fw<-unlist(import.bw(bigWigFiles[i], as = "NumericList")[[1]])
  x.rv<-unlist(import.bw(bigWigFiles.rv[i], as = "NumericList")[[1]])
  x<-c(x.fw, x.rv)
  y<-vector()
  for(j in 1:nrow(TSS.called.final)){
    if(TSS.called.final$strand[j]=="+")
      z<-sum(x[(TSS.called.final$genomePos[j]-5):(TSS.called.final$genomePos[j]+5)])
    
    if(TSS.called.final$strand[j]=="-")
    
      z<-sum(x[(TSS.called.final$genomePos[j]-5+2522992):(TSS.called.final$genomePos[j]+5+2522992)])
    
    y[j]<-z
  }
  countData[,i]<-y
}
  
condition <- factor(c("wt","wt","ko","ko"))
    
dds <- DESeqDataSetFromMatrix(countData, DataFrame(condition), ~ condition)
dds <- DESeq(dds)
deseq.results<-results(dds)

plotMA(deseq.results, ylim=c(-10,10))

TSS.called.final[deseq.results$log2FoldChange >2,]
```

```{r volcano plot}
deseq.results.df <-data.frame(baseMean = deseq.results$baseMean,
                              log2FoldChange = deseq.results$log2FoldChange,
                              padj = deseq.results$padj)

crispr.filt <-  TSS.called.final$genomePos %in% crispr_loci[1,2]:crispr_loci[1,3] |
                TSS.called.final$genomePos %in% crispr_loci[2,2]:crispr_loci[2,3]


nonCRISPR.filt <- vector(length = nrow(TSS.called.final))
for (i in 1: nrow(TSS.called.final)){
  x <- min(abs(TSS.called.final$genomePos[i] -  nonCRISPR$V3)) <= 50
  nonCRISPR.filt[i] <- x
}
summary(nonCRISPR.filt) 

leader.filt <- TSS.called.final$genomePos %in% crispr_leader[1,2]:crispr_leader[1,3] |
               TSS.called.final$genomePos %in% crispr_leader[2,2]:crispr_leader[2,3]


#volcano plot

#set limit for -log10 padj
p.lim = 60

g <- ggplot(data=deseq.results.df[!crispr.filt & !nonCRISPR.filt,]) + 
     geom_hline(aes(yintercept = 2), show.legend = F, lty = 2) +
     geom_point(aes(x=log2FoldChange, y= ifelse(-log(padj,10) < p.lim, -log(padj,10), p.lim), shape = ifelse(-log(padj,10) < p.lim, "16", "17")), cex=1, alpha=0.05)
g <- g  + geom_point(data=deseq.results.df[nonCRISPR.filt,] ,    aes(x=log2FoldChange, y= ifelse(-log(padj,10) < p.lim, -log(padj,10), p.lim), shape = ifelse(-log(padj,10) < p.lim, "16", "17")), colour="#1289A7", cex=1, alpha=0.8)
g <- g  + geom_point(data=deseq.results.df[crispr.filt,] , 
                     aes(x=log2FoldChange, y= ifelse(-log(padj,10) < p.lim, -log(padj,10), p.lim), shape = ifelse(-log(padj,10) < p.lim, "16", "17")), colour="#FFC312", cex=1, alpha=0.8)
g <- g + theme_bw() + theme(aspect.ratio=1, panel.grid = element_blank()) +
  scale_x_continuous(limits=c(-11,11), breaks=seq(-10, 10,by=2), expand = c(0, 0)) +
  scale_y_continuous(limits=c(-1, p.lim+1), breaks=seq(0, 120, by=20), expand = c(0, 0)) +
  ylab("log10(padj)") +
  xlab("log2-fold change") +
  theme(legend.position='none')
g
ggplot2::ggsave("../plots/TSS.deseq.png", plot = g, device = "png", path = NULL,
       scale = 1, width = 80, height = 80, units = c("mm"), limitsize = TRUE)
```



```{r Fisher test categories}
sl <- 0.01
#checking fraction of each category with significant changes
summary(deseq.results$padj < sl)
summary(deseq.results$padj[!crispr.filt & !nonCRISPR.filt] < sl)
summary(deseq.results$padj[nonCRISPR.filt] < sl)
summary(deseq.results$padj[crispr.filt] < sl)

# fraction of genes called to be differentially regulated (padj < 0.01)
as.numeric(summary(deseq.results$padj < sl)[3])/length(deseq.results$padj)

#Fisher's exact test for enrichment of upregulated TSS RNAs upon cbp1 deletion
up.m<-matrix(c(sum(deseq.results$padj[crispr.filt & deseq.results$log2FoldChange < 0] < sl),
                     sum(crispr.filt) - sum(deseq.results$padj[crispr.filt & deseq.results$log2FoldChange < 0] < sl),
                     sum(deseq.results$padj[nonCRISPR.filt & deseq.results$log2FoldChange < 0] < sl),
                     sum(nonCRISPR.filt) - sum(deseq.results$padj[nonCRISPR.filt & deseq.results$log2FoldChange < 0] < sl),
                     sum(deseq.results$padj[!crispr.filt & !nonCRISPR.filt & deseq.results$log2FoldChange < 0] < sl),
                     sum(!crispr.filt & !nonCRISPR.filt) - sum(deseq.results$padj[!crispr.filt & !nonCRISPR.filt & deseq.results$log2FoldChange < 0] <  sl)),
                     nrow=2)
up.m
p.CRISPR.up<-fisher.test(up.m[,c(1,3)])$p.value
p.nonCRISPR.up<-fisher.test(up.m[,c(2,3)])$p.value


#Fisher's exact test for enrichment of downregulated TSS RNAs upon cbp1 deletion
down.m <- matrix(c(sum(deseq.results$padj[crispr.filt & deseq.results$log2FoldChange > 0] < sl),
                     sum(crispr.filt) - sum(deseq.results$padj[crispr.filt & deseq.results$log2FoldChange > 0] <sl),
                     sum(deseq.results$padj[nonCRISPR.filt & deseq.results$log2FoldChange > 0] < 0.01),
                     sum(nonCRISPR.filt) - sum(deseq.results$padj[nonCRISPR.filt & deseq.results$log2FoldChange > 0] < sl),
                     sum(deseq.results$padj[!crispr.filt & !nonCRISPR.filt & deseq.results$log2FoldChange > 0] < sl),
                     sum(!crispr.filt & !nonCRISPR.filt) - sum(deseq.results$padj[!crispr.filt & !nonCRISPR.filt & deseq.results$log2FoldChange > 0] < sl)),
                     nrow=2)
down.m
p.CRISPR.down<-fisher.test(down.m[,c(1,3)])$p.value
p.nonCRISPR.down<-fisher.test(down.m[,c(2,3)])$p.value

#manual bonferroni correction
p.vec<-c(p.CRISPR.up, p.nonCRISPR.up, p.CRISPR.down, p.nonCRISPR.down)
p.adjust(p.vec, method="bonferroni")
```

```{r identify TSSs regulated by Cbp1}
sl <- 0.01
deseq.results.df[nonCRISPR.filt,][deseq.results.df[nonCRISPR.filt,]$padj < sl,]
TSS.called.final[nonCRISPR.filt,][deseq.results.df[nonCRISPR.filt,]$padj < sl,]

#exporting those TSSs upregulated in cbp1 deletion strain
ex <- TSS.called.final[nonCRISPR.filt,][deseq.results.df[nonCRISPR.filt,]$padj < sl & deseq.results.df[nonCRISPR.filt,]$log2FoldChange < 0,]
data.bed <- data.frame(rep(sisl.genome[1,1], times=nrow(ex)), ex$genomePos-1, ex$genomePos, rep("TSS", times=nrow(ex)), rep("*", times=nrow(ex)), ex$strand)
write.table(data.bed, "../data/Cappable-seq_S.islandicus/TSSs.Cbp1Associated.upregulatedInCbp1DeletionStrain.bed", quote=F, row.names=F, col.names=F, sep = "\t")
```



```{r identify TSSs regulated by Cbp1 in CRISPR arrays}

deseq.results.df[crispr.filt,][deseq.results.df[crispr.filt,]$padj < sl,]
TSS.called.final[crispr.filt,][deseq.results.df[crispr.filt,]$padj < sl,]

```

Plotting fractions of significantly up- and down-regulated genes for each category
```{r plot fraction regulated}
df <-data.frame(abs=c(up.m[1,], down.m[1,]), frac=c(up.m[1,]/colSums(up.m),down.m[1,]/colSums(down.m)), type=rep(c("CRISPR arrays", "non-CRISPR binding sites", "not Cbp1 associated"),2), dir=c(rep("increase",3), rep("decrease",3)))
g1 <- ggplot(data=df, aes(x=dir, y=frac, group=type)) + geom_col(aes(fill=type), position="dodge")
g1 <- g1 + geom_text(aes(label=abs), position=position_dodge(0.9), vjust=-0.5) #add absolute counts for each group as label
g1 <- g1 + scale_y_continuous("fraction", limits=c(0,1), expand=c(0,0)) + theme_bw() + theme(aspect.ratio=1, panel.grid = element_blank())
g1 <- g1 + scale_fill_manual(values=c("#afdde9", "#2c89a0", "#164450")) + xlab("Relative abundance of TSS RNAs in deltaCbp1")
g1

ggsave("../plots/TSS.deseq.bargraph.pdf", plot = g1, path = NULL,
       scale = 1, width = 160, height = 160, units = c("mm"), limitsize = TRUE)
```

```{r QC scatter plots for replicates}
cor(countData, method="spearman")

par(mfrow=c(3,3), mar=c(5,4,2,2))
for(i in 2:4){
  for(j in 1:3){
    out <- heatscatter(countData[,i], countData[,j], log="xy", xlab="", ylab="", main="", xlim=c(1,10^6), ylim=c(1,10^6), cex=0.2)
  }
}




```

```{r circos plot}
#make data frame for circos plot
sl <- 0.01

dbed<-data.frame(chr=rep("NC_017276.1", times=nrow(deseq.results)), start= TSS.called.final$genomePos -1, end= TSS.called.final$genomePos)
dbed$logFC = - deseq.results$log2FoldChange
dbed$padj <- deseq.results$padj < sl

#circos plot
pdf(file=paste0("../plots/circos.cappables-seq.padj", sl,".pdf"), width=4, height=4)  
circos.clear()
  circos.par("track.height" = 0.6, start.degree = 90, gap.degree = 0, cell.padding = c(0.0, 0,0.0, 0))
  circos.initializeCircularGenome(name="NC_017276.1", genome_size=2522992, major.by=200000)
  draw.sector(338.9, 348, rou1 = 1, rou2 = 0.1, clock.wise = F, col = "#ff000010", border=F) 
#Track showing genomic positions of CIDs
  circos.genomicTrack(data=cid, ylim=c(0,1), track.height=0.04, panel.fun = function(region, value, ...) {
      circos.genomicRect(region, value, ybottom = 0, col = "lightgrey", border = "white")
          }, bg.border = NA)
#Track showing genomic positions of CRISPR arrays 
 circos.genomicTrack(data=crispr, ylim=c(0,1), track.height=0.04, panel.fun = function(region, value, ...) {
    circos.genomicRect(region, value, ybottom = 0, col = "#afdde9ff", border = "#afdde9ff")
        }, bg.border = NA)
#Track showing log2FC of genes with color based on padj  
  circos.genomicTrack(data=dbed, numeric.column=4, track.height=0.64, panel.fun = function(region, value, ...) {
      circos.genomicPoints(region, value, pch=16, cex=0.4,
              col = ifelse(value[[2]] == T & value[[1]] > 0, add_transparency("cyan4", transparency=0.4),
                                  ifelse(value[[2]] == T & value[[1]] < 0, add_transparency("red", transparency=0.4),
                                        add_transparency("grey", transparency = 0.6))), ...)
             }, bg.border = NA)
       circos.segments(x0=0, y0=0, x1=2522992, y1=0)
  circos.yaxis("right", labels.niceFacing = F)
  dev.off()

```

```{r assigning primary TSSs}

operons2 <- data.frame(PosLeft=vector(), PosRight=vector(), strand=vector(), type=vector(), first.cistron=vector())
for(i in 1:length(unique(operons$Operon))){
  subs <- operons[operons$Operon == i,]
  x <- c(min(subs$PosLeft), 
         max(subs$PosRight),
         subs$Strand[1], 
         subs[1,]$Type, 
         ifelse(subs$Strand[1] == "+", subs$IdGene[1], subs$IdGene[nrow(subs)]))
  operons2 <- rbind(operons2, x)
}

colnames(operons2) <- c("PosLeft", "PosRight", "Strand", "Type", "First.cistron")
operons2$PosLeft <- as.numeric(operons2$PosLeft)
operons2$PosRight <- as.numeric(operons2$PosRight)
head(operons2)



#calculate upstream intergenic region for each operon
md <- 64 #maximum distance allowed for upstream IGR search window, with fraction of  0.98 of TSSs in Wurtzel et al 2010 Genome Research 
start.igr <- vector()
end.igr <- vector()

for(i in 1:nrow(operons2)){
  s <- operons2$Strand[i]
  if(s == "+"){
    nd <- operons2$PosLeft[i] 
    strt <- ifelse(i>1, max(nd-md, operons2$PosRight[i-1]), operons2$PosLeft[i]-md)
     }else{
    strt <- operons2$PosRight[i]
    nd <- ifelse(i<nrow(operons2), min(strt+md, operons2$PosLeft[i+1]), operons2$PosRight[i] + md)  
  }
  start.igr[i] <- strt
  end.igr[i] <- nd
}
igr <- data.frame(start=start.igr, end=end.igr, strand=operons2$Strand)

#take TSS with maximum intensity in IGR with additional requirement for max distance (md) from start codon

prim <- vector() #vector recording primary TSS position
dist <- vector() #vector recording distance of primary TSS to start codon, i.e. 5'UTR length
for(i in 1:nrow(igr)){
 w  <- which(TSS.called.final$strand == igr$strand[i] & dplyr::between(TSS.called.final$genomePos, igr$start[i], igr$end[i]))
  v <- TSS.called.final[w,]
  u <- countData[w,]
  if(length(u)>=8){t <- which.max(u[,1]*u[,2])
  s <- v[t,1]}else{
    if(length(u)==4){s <- v[1,1]}else{
      s <- NA}
  }
  prim[i] <- s
  if(igr$strand[i] == "+"){
    r <- igr$end[i] - s}else{
      r <- s - igr$start[i]
    }
  dist[i] <- r
}

#number of primary TSSs
summary(is.na(prim))

#number of coding operons
summary(operons2$Type == "CDS")

#number of coding operons with primary TSSs assigned
sum(!is.na(prim) & operons2$Type == "CDS")

hist(dist[!is.na(prim) & operons2$Type == "CDS"], breaks = md)

#calculate fraction of 5'UTRs <=3 nt, 
sum(!is.na(dist) & operons2$Type == "CDS" & dist <= 3)/sum(!is.na(dist) & operons2$Type == "CDS")


#ggplot histogram
dist2 <- data.frame(dist=dist[!is.na(dist) & operons2$Type == "CDS"])
#dist2[dist2 > 20] <- 21

gh <- ggplot(data = dist2, aes(x=dist)) + geom_bar()
gh <- gh + scale_y_continuous("Frequency", limits=c(0,300), expand=c(0,0)) 
gh <- gh + scale_x_continuous(limits=c(-1,65), breaks=seq(0,64, by=2), expand=c(0,0)) 
gh <- gh + theme_bw() + theme(aspect.ratio=0.6, panel.grid = element_blank())
gh <- gh + scale_fill_manual(values=c("#164450")) + xlab("5' UTR length (nt)")
gh
ggsave("../plots/5UTRlegth.bargraph.pdf", plot = gh, path = NULL,
       scale = 1, width = 160, height = 80, units = c("mm"), limitsize = TRUE)
```

```{r exporting primary TSS data}
#export of primary TSS data for coding genes and the two CRISPR arrays, log2-fold change and padj and first downstream cistron to compare to RNA-seq results

primTSS.df <- data.frame(pos = prim[!is.na(prim) & operons2$Type == "CDS"], 
                         strand = operons2$Strand[!is.na(prim) & operons2$Type == "CDS"],
                         first.cistron = operons2$First.cistron[!is.na(prim) & operons2$Type == "CDS"])


#identify TSSs of CRISPR arrays, searching 50 nt upstream of CRISPR arrays
crisprTSS <- vector()
for(i in 1:2){
  d <- sisl.genome[sisl.genome$V3 == "repeat_region",]
  crisprTSS[i] <- TSS.called.final$genomePos[between(TSS.called.final$genomePos,
                                                     ifelse(d$V7[i] == "+", d$V4[i] - 40, d$V5[i]),
                                                     ifelse(d$V7[i] == "+", d$V4[i], d$V5[i] + 40))]
}

crisprTSS.df <- data.frame(pos = crisprTSS, 
                           strand = sisl.genome[sisl.genome$V3 == "repeat_region", 7],
                           first.cistron = c("id2", "id3"))
primTSS.df <- rbind(primTSS.df, crisprTSS.df)

#merging deseq.results
sel.v <- vector()
for (i in 1:nrow(primTSS.df)){
  c <- which(TSS.called.final$genomePos == primTSS.df$pos[i]  & TSS.called.final$strand == primTSS.df$strand[i])
  sel.v[i] <- c
}
primTSS.df <- cbind(primTSS.df, deseq.results.df[sel.v,])

write.table(primTSS.df, "../data/Cappable-seq_S.islandicus/deseq.primaryTSSs.txt", col.names=T, row.names=F, quote=F, sep="\t")
```


```{r count internal sense and antisense transcripts}
#filter gff file for gene entries
sisl.genome2 <- sisl.genome %>% dplyr::filter(V3 == "gene")

as.count <- vector()
sense.count <- vector()

for(i in 1:nrow(sisl.genome2)){
  s <- sisl.genome2$V7[i]
  a <-sum(dplyr::between(TSS.called.final$genomePos , sisl.genome2$V4[i], sisl.genome2$V5[i]) & TSS.called.final$strand != s)
  b <-sum(dplyr::between(TSS.called.final$genomePos , 
                         ifelse(s == "+",  sisl.genome2$V4[i] +1, sisl.genome2$V4[i]), 
                         ifelse(s == "+",  sisl.genome2$V5[i], sisl.genome2$V5[i] -1)) & 
                         TSS.called.final$strand == s)
  as.count[i-1] <- a 
  sense.count[i-1] <- b
}

sum(as.count) #number of antisense internal TSSs
sum(sense.count) #number of sense internal TSSs
```



